### PR TITLE
Feature/migration from url - closes #6

### LINKED ISSUE #6: Implement migration and database creation startup flags
Implement a command line flag `--with-migration <url>` that will instruct the main program to download data from a URL and insert in the database before bringing the server up.

If the database already has data, throw an error. We could create another flag `--recreate-database` that will drop and create a database from scratch. So, if the user executes:
```
$ stop-analyzing-api --recreate-database --with-migration https://endpoint-with-migration-data
```

The service will perform a clean start loading this data. This will be very useful for development porpuses.

How about that?

@jairsjunior or @flaviostutz, any thoughts?

### PR DESCRIPTION
**Pull request by zamariola**
This PR is a work in progress; however, considering that we have two issues #6 and #7 related to the same flow, I decided to create this PR, so we can work together.

I tried to follow as much as possible the current structure but I have two possible suggestions:
1) Change the migration from a plain go file containing URL and File to maybe one interface and two migration structs:

```
type Migrator interface {
   func MIgrate(path string) error
}

type URLMigrator struct

type FileMigrator struct
```

2) Inject database instance into those structs to improve testability instead of getting it as a global var.

What do you guys think?

### CODE REVIEW

**Comment by tiagostutz** [comment]
Hi @zamariola ! Nice you have opened the PR...! 👍 

I like the idea... So, how would we decie whether this is a **file** or **URL** migrator? Will it be based on the `path` pattern (e.g. if it has `http[s]`)?

About the Database injection, maybe we could have another interface to handle the database operations and this interface here uses it instead of using directly the Database, so other interfaces that needs to deal with data uses this Database structure, something like the old _DAOs_ or _Repositories_. 

This Database handling interface would then get Database configurations directly from args (flags). 

What do you think?

**Comment by eniovalo** [comment]
@zamariola and @tiagostutz , great idea. =]

About the database handling, it would be something like *TopUserGetter Interface* in this [article](https://github.com/Evertras/go-interface-examples/tree/master/local-interfaces#first-stop-leaderboard) ?

**Comment by zamariola** [comment]
> @zamariola and @tiagostutz , great idea. =]
> 
> About the database handling, it would be something like _TopUserGetter Interface_ in this [article](https://github.com/Evertras/go-interface-examples/tree/master/local-interfaces#first-stop-leaderboard) ?

Exactly! Awesome blog post, btw.

**Comment by zamariola** [comment]
Hello @tiagostutz I think we can still use the isURL accessory method to get the correct interface implementation. It'd be something like.
```
if isURL(path) {
     URLMigrator{dbInstance}.migrate(path)
}

FileSystemMigrator{dbInstance}.migrate(path)


```
In this case, I think I can create both structs and the interface in the same file (my old java soul tends to force me to create a new file to every struct though 🤣 )
---

About the database instance, I don't have any problems with our current  ` db.Connect()`. It was more a suggestion about how to use it inside the migrators, so we can mock it in an easier way. I'm going to create a commit using it and then we can double-check it together.

Thanks!

**Comment by tiagostutz** [comment]
> In this case, I think I can create both structs and the interface in the same file (my old java soul tends to force me to create a new file to every struct though 🤣 )

Do you know what would be a Java thing? Check this:
```
migrator := MigratorFactory.create(path)
migrator.migrate()
```
God damn it! ahahahahaha...


Great suggestions, lets discuss the implementation as it goes.

**Comment by eniovalo** [comment]
@zamariola and @tiagostutz ,
I read a little about Interfaces in Go in this [article](https://medium.com/rungo/interfaces-in-go-ab1601159b3a) and tried to refact *migration.go* to use interface.

```go
package migration

import (
	"encoding/json"
	"fmt"
	"io/ioutil"
	"net/http"
	"net/url"
	"time"

	"github.com/bancodobrasil/stop-analyzing-api/db"

	"github.com/sirupsen/logrus"
)

//Item represents a migration (file and url) entry point for an Item entity
type Item struct {
	Title      string   `json:"title"`
	Subtitle   string   `json:"subtitle"`
	ContentURL string   `json:"contentURL"`
	Tags       []string `json:"tags"`
}

//Migrater is an interface to migrate from URL or JSon File.
type Migrater interface {
	Migrate() error
}

//URLMigrater is a struct to migrate from URL.
type URLMigrater struct {
	dbInstance db.DatabasePrisma
	url        string
	recreate   bool
}

//FileMigrater is a struct to migrate from JSon file.
type FileMigrater struct {
	dbInstance db.DatabasePrisma
	filePath   string
	recreate   bool
}

/*
New initialize a type of Migrate Struct depend on the path received.

Return URLMigrater if the path is an URL or FileMigrater if the path is a Json File.
*/
func New(path string, recreate bool) (Migrater, error) {
	dbInstance, err := db.Connect()
	if err != nil {
		return nil, err
	}

	//If something happens, disconnect database in the end.
	defer dbInstance.Disconnect()

	if isURL(path) {
		return &URLMigrater{dbInstance, path, recreate}, nil
	}

	return &FileMigrater{dbInstance, path, recreate}, nil
}

/*
Migrate function migrate data from an URL.
*/
func (u *URLMigrater) Migrate() error {
	logrus.Infof("Migrating database from URL: %s", u.url)

	body, err := executeGET(u.url)
	if err != nil {
		return err
	}

	items, err := parseItems(body)
	if err != nil {
		return err
	}

	return importItems(items)
}

/*
Migrate function migrate data from a json file.
*/
func (f *FileMigrater) Migrate() error {
	logrus.Infof("Migrating database from file: %s", f.filePath)

	jsonFile, err := ioutil.ReadFile(f.filePath)
	if err != nil {
		logrus.Errorf("Occured an error while reading file: %s", err)
	}

	items, err := parseItems(jsonFile)
	if err != nil {
		return err
	}

	logrus.Infof("file parsed: %s", items)

	return importItems(items)
}

func parseItems(body []byte) ([]Item, error) {

	items := make([]Item, 0)

	if err := json.Unmarshal(body, &items); err != nil {
		return nil, err
	}

	return items, nil
}

func importItems([]Item) error {
	//TODO: Import items
	return nil
}

func isURL(str string) bool {
	u, err := url.Parse(str)
	return err == nil && u.Scheme != "" && u.Host != ""
}

func executeGET(url string) ([]byte, error) {

	cli := http.Client{Timeout: 30 * time.Second}
	resp, err := cli.Get(url)

	if err != nil {
		return nil, err
	}

	if resp.StatusCode != http.StatusOK {
		return nil, fmt.Errorf("Unable to get value, unexpected response status code %d from %s", resp.StatusCode, url)
	}

	return ioutil.ReadAll(resp.Body)
}
```

And I change the *cmd/migration.go* to work with this.

```go
RunE: func(cmd *cobra.Command, args []string) error {
	recreateDB := viper.GetViper().GetBool("recreate-database")

	migrater, err := migration.New(args[0], recreateDB)
	if err != nil {
		return err
	}

	return migrater.Migrate()
},
```

What do you think? :slightly_smiling_face:

**Comment by tiagostutz** [comment]
Hi @eniovalo ! Looks good for me.!

There's just two things to point out:
1) There's a typo on the name "Migrater". Change it to "Migrator", please.
2) I totally forgot in our discussion that we won't make database handling direct; instead, we are using [Prisma.io](https://github.com/prisma/prisma-client-go/blob/master/docs/quickstart.md) for this database layer. So, the function `importItems` will use the `*db.DatabasePrisma` to access the database. Check the README for instructions and the file **service/v1/routes.go** for an simple example of how to retrieve tags from database.

**Comment by zamariola** [comment]
Thank for the input @eniovalo I'm going to use that structure changing the importItems with the db as tiago mentioned. :+1:

**Comment by zamariola** [comment]
I'm planning to work on the real data import now:
- search/import tags
- import image refs
- create an item table to store the option (?)

**Comment by eniovalo** [comment]
Hi @tiagostutz ,

> 1. There's a typo on the name "Migrater". Change it to "Migrator", please.

Thanks, I tried to follow [effective_go article](https://golang.org/doc/effective_go.html#interface-names) to plus a sufix **er** , but it got strange. hehe

> 2. I totally forgot in our discussion that we won't make database handling direct; instead, we are using [Prisma.io](https://github.com/prisma/prisma-client-go/blob/master/docs/quickstart.md) for this database layer. So, the function `importItems` will use the `*db.DatabasePrisma` to access the database. Check the README for instructions and the file **service/v1/routes.go** for an simple example of how to retrieve tags from database.

Thanks, I was reading [Prisma API Docs](https://github.com/prisma/prisma-client-go/blob/master/docs/reference/basic.md) to understand the functions. 🙂

**Comment by zamariola** [comment]
I think we have a reviewable version now @eniovalo @tiagostutz

We are dropping tags and items (if needed) and then recreating them.

I stumped upon a challenge. I couldn't find a way to create an Item with two+ tags attached. So we need to iterate on tags and link them one-by-one. :thinking: 

Thanks!

**Comment by tiagostutz** [comment]
Really good news, @zamariola !

I'm going to review it right now.! =)

File: src/db/database_integration_test.go [hunk 1]
```diff
@@ -0,0 +1,185 @@
...
+
+	fmt.Println(fetchedItem.Tags())
+
+	if len(fetchedItem.Tags()) != 2 {
+		t.Errorf("Wrong item tags, expected size: %d, found: %d", 2, len(fetchedItem.Tags()))
+	}
+}
+
+func configEnvVars() {
+	os.Setenv("DATABASE_URL", "postgresql://user2020:pass2020@localhost:5432/stop-analyzing-api")
```
Comment (tiagostutz) [inline src/db/database_integration_test.go:184]: I think we could replace the use of env vars in the Go program by flags. Then when we start the programa we use env vars to set those flag values... I think it is a more direct approach.

This is just a comment, don't need to work on it on this PR, ok? Everthing is perfect here.

File: src/db/database.go [hunk 2]
```diff
@@ -33,40 +33,126 @@ func (d *DatabasePrisma) Disconnect() {
...
+	if err != nil {
+		return item, err
+	}
+
+	//Link tags - TODO: Check if its possible to use/create one method to link them all
+	findResult := d.client.Item.FindOne(
+		Item.ID.Equals(item.ID),
+	)
+	for _, mTag := range mTags {
+		item, err = findResult.Update(
```
Comment (eniovalo) [inline src/db/database.go:135]: I didn't test it, but I think *err* variable would have the last error occurred at findResult.Update(), so if occurred 3 errors, we would only return the last one. By the way, I don't see much problems in this situation. :slightly_smiling_face:

**Comment by tiagostutz** [comment]
@eniovalo @zamariola  I just tested local here and for me this PR is good to go! 🏁 🏎️  

@zamariola, really good job on this PR dude!

File: src/db/database.go (hunk 2 above)
Comment (zamariola) [reply src/db/database.go:135]: Good point! I don't foresee anything that could break it, but I think we can raise a new issue about it, what do you think?

**Comment by zamariola** [comment]
Aewsome!

File: src/db/database.go (hunk 2 above)
Comment (eniovalo) [reply src/db/database.go:135]: Yes, good idea, we can create a new issue for this situation. =]

**Comment by tiagostutz** [comment]
@zamariola Your PR is titled with [ WIP ] and I wasn't sure whether we should or not merge this. Do you have more stuff to add or can we merge this right now?

**Comment by zamariola** [comment]
Sure! my bad! We can merge it! Thanks!

**Comment by tiagostutz** [comment]
@zamariola thanks for this great PR! I just sent you and invitation to be part of the project team. If you could accept it would be of great value for the project.

Maybe you take a look at this discussion in #11 and, if you want, share some thoughts there.

Thanks again for your valuable contribution and for your engagement with the project. I think we will release a simple but great tool out there. 🤝

**Comment by zamariola** [comment]
Sure @tiagostutz ! I've just accepted the invitation. Thank you for it! I'm going to go through all open issues to double-check how I can give some 20 cents here and there.

**Pull request merged**